donazioni/viste.py

```diff
@@ -181,7 +187,73 @@ def etichette_elenco(request, me):
     campagne_qs = me.oggetti_permesso(GESTIONE_CAMPAGNA)
     filtro_etichette = Q(comitato__in=chain(comitati, sedi_deleghe_campagne)) | Q(campagne__in=campagne_qs)
     contesto = {
-        "etichette": Etichetta.objects.filter(filtro_etichette).distinct(),
+        'etichette': Etichetta.objects.filter(filtro_etichette).distinct(),
         'puo_modificare': me.ha_permessi(GESTIONE_CAMPAGNE)
     }
     return 'donazioni_etichette_elenco.html', contesto
+
+
+# ############# DONAZIONI e DONATORI ###################
+
+@pagina_privata
+def donazione_nuova(request, me, campagna_id):
+    campagna = get_object_or_404(Campagna, pk=campagna_id)
+    if campagna not in me.oggetti_permesso(GESTIONE_CAMPAGNA):
+        return redirect(ERRORE_PERMESSI)
+
+    modulo_donazione = ModuloDonazione(request.POST or None, campagna=campagna_id)
+    modulo_donatore = ModuloDonatore(request.POST or None)
+    if modulo_donazione.is_valid():
+        donazione = modulo_donazione.save(commit=False)
+
+        if modulo_donatore.is_valid():
+            donatore = Donatore.nuovo_o_esistente(modulo_donatore.instance)
+            donazione.donatore = donatore
+
+        donazione.save()
+        return redirect(reverse('donazioni_campagna', args=(campagna_id,)))
+    contesto = {
+        'modulo_donazione': modulo_donazione,
+        'modulo_donatore': modulo_donatore,
+        'campagna': campagna,
+    }
+    return 'donazioni_donazione_nuova.html', contesto
+
+
+@pagina_privata
+def donazioni_elenco(request, me, campagna_id):
+    campagna = get_object_or_404(Campagna.objects.prefetch_related('donazioni', 'donazioni__donatore'), pk=campagna_id)
+    if campagna not in me.oggetti_permesso(GESTIONE_CAMPAGNA):
+        return redirect(ERRORE_PERMESSI)
+    totale_donazioni = campagna.donazioni.all().aggregate(importo=Sum('importo'))
+    numero_donatori_censiti = campagna.donazioni.filter(donatore__isnull=False).distinct('donatore').count()
+
+    contesto = {
+        'campagna': campagna,
+        'totale_donazioni': totale_donazioni['importo'],
+        'numero_donatori_censiti': numero_donatori_censiti,
+    }
+    return 'donazioni_campagna_elenco_donazioni.html', contesto
+
+
+@pagina_privata
+def donatori_campagna_elenco(request, me, campagna_id):
+    campagna = get_object_or_404(Campagna, pk=campagna_id)
+    if campagna not in me.oggetti_permesso(GESTIONE_CAMPAGNA):
```

Comment:
No - questo non e' efficiente. `oggetti_permesso` puo' ritornare un QuerySet molto largo e questa operazione (operatore `in`) ha complessita' lineare. 

Ti prego di rileggere la documentazione relativa ai permessi sul Wiki. In particolare, per controllare i permessi su un particolare oggetto, usa `me.permessi_almeno(oggetto, livello)`. Questo metodo e' altamente ottimizzato, evita molte query inutili e quando le fa, usa operazioni piu' piccole (es. `.exists()` invece di ciclare su tutti i risultati).